### PR TITLE
gRPC reflection security

### PR DESCRIPTION
**Pull request by JamesNK**
Addresses https://github.com/grpc/grpc-dotnet/issues/2483

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/grpc/test-tools.md](https://github.com/dotnet/AspNetCore.Docs/blob/3df317a9a7729a9214a7b51e2a2c6d0772829d3d/aspnetcore/grpc/test-tools.md) | [Test gRPC services with gRPCurl and gRPCui in ASP.NET Core](https://review.learn.microsoft.com/en-us/aspnet/core/grpc/test-tools?branch=pr-en-us-33092) |

### CODE REVIEW

**Comment by JamesNK** [comment]
Could an expert fix up the busted API links.

File: aspnetcore/grpc/test-tools.md [hunk 1]
```diff
@@ -50,6 +50,30 @@ When gRPC reflection is set up:
...
+gRPC reflection is often only required in a local development environment. In this situation, the service should only be mapped when <xref:Microsoft.AspNetCore.Hosting.HostingEnvironmentExtensions.IsDevelopment(IHostingEnvironment)> returns true:
+
+```csharp
+if (env.IsDevelopment())
+{
+    app.MapGrpcReflectionService();
+}
+```
+
+Access to the service can be controlled through standard ASP.NET Core authorization extension methods, such as <xref:Microsoft.AspNetCore.Builder.AuthorizationEndpointConventionBuilderExtensions.AllowAnonymous<TBuilder>(TBuilder)> and <xref:Microsoft.AspNetCore.Builder.AuthorizationEndpointConventionBuilderExtensions.RequireAuthorization<TBuilder>(TBuilder)>.
```
Comment (Rick-Anderson) [inline aspnetcore/grpc/test-tools.md:66]:
```suggestion
Access to the service can be controlled through standard ASP.NET Core authorization extension methods, such as [`AllowAnonymous<TBuilder>`](/dotnet/api/microsoft.aspnetcore.builder.authorizationendpointconventionbuilderextensions.allowanonymous) and [`RequireAuthorization<TBuilder>`](/dotnet/api/microsoft.aspnetcore.builder.authorizationendpointconventionbuilderextensions.requireauthorization).
```

File: aspnetcore/grpc/test-tools.md [hunk 2]
```diff
@@ -50,6 +50,30 @@ When gRPC reflection is set up:
...
+
+For example, if an app has been configured to require authorizaiton by default, configuration the gRPC reflection endpoint with `AllowAnonymous()` to skip authentication and authorization.
```
Comment (Rick-Anderson) [inline aspnetcore/grpc/test-tools.md:68]:
```suggestion
For example, if an app has been configured to require authorization by default, configuration the gRPC reflection endpoint with `AllowAnonymous` to skip authentication and authorization.
```

File: aspnetcore/grpc/test-tools.md [hunk 3]
```diff
@@ -50,6 +50,30 @@ When gRPC reflection is set up:
 * Client apps that support gRPC reflection can call the reflection service to discover services hosted by the server.
 * gRPC services are still called from the client. Reflection only enables service discovery and doesn't bypass server-side security. Endpoints protected by [authentication and authorization](xref:grpc/authn-and-authz) require the caller to pass credentials for the endpoint to be called successfully.
 
+### Reflection service security
+
+gRPC reflection returns a list of available APIs, which could be sensitive information. Care should be taken to limit access to the gRPC reflection service.
```
Comment (Rick-Anderson) [inline aspnetcore/grpc/test-tools.md:55]:
```suggestion
gRPC reflection returns a list of available APIs, which could contain sensitive information. Care should be taken to limit access to the gRPC reflection service.
```

File: aspnetcore/grpc/test-tools.md [hunk 4]
```diff
@@ -50,6 +50,30 @@ When gRPC reflection is set up:
...
+
+gRPC reflection is often only required in a local development environment. In this situation, the service should only be mapped when <xref:Microsoft.AspNetCore.Hosting.HostingEnvironmentExtensions.IsDevelopment(IHostingEnvironment)> returns true:
```
Comment (Rick-Anderson) [inline aspnetcore/grpc/test-tools.md:57]:
```suggestion
gRPC reflection is usually only required in a local development environment. For local development, the reflection service should only be mapped when <xref:Microsoft.AspNetCore.Hosting.HostingEnvironmentExtensions.IsDevelopment(IHostingEnvironment)> returns true:
```

Comment (wadepickett) [inline aspnetcore/grpc/test-tools.md:57]:
```suggestion
gRPC reflection is often only required in a local development environment. In this situation, the service should only be mapped when <xref:Microsoft.Aspnetcore.Hosting.HostingEnvironmentExtensions.Isdevelopment> returns true:
```

Comment (Rick-Anderson) [inline aspnetcore/grpc/test-tools.md:57]:
```suggestion
gRPC reflection is often only required in a local development environment. In this situation, the service should only be mapped when [IsDevelopment](/en-us/dotnet/api/microsoft.aspnetcore.hosting.hostingenvironmentextensions.isdevelopment) returns true:
```

**Review by Rick-Anderson — CHANGES_REQUESTED** [review]

File: aspnetcore/grpc/test-tools.md (hunk 4 above)
Comment (wadepickett) [reply aspnetcore/grpc/test-tools.md]: Looks like Rick and I are in at same time, I will get out.

File: aspnetcore/grpc/test-tools.md [hunk 5]
```diff
@@ -50,6 +50,30 @@ When gRPC reflection is set up:
 * Client apps that support gRPC reflection can call the reflection service to discover services hosted by the server.
 * gRPC services are still called from the client. Reflection only enables service discovery and doesn't bypass server-side security. Endpoints protected by [authentication and authorization](xref:grpc/authn-and-authz) require the caller to pass credentials for the endpoint to be called successfully.
 
+### Reflection service security
+
+gRPC reflection returns a list of available APIs, which could contain sensitive information. Care should be taken to limit access to the gRPC reflection service.
+
+gRPC reflection is often only required in a local development environment. In this situation, the service should only be mapped when <xref:Microsoft.Aspnetcore.Hosting.HostingEnvironmentExtensions.Isdevelopment> returns true:
```
Comment (JamesNK) [inline aspnetcore/grpc/test-tools.md:57]:
```suggestion
gRPC reflection is usually only required in a local development environment. For local development, the reflection service should only be mapped when [IsDevelopment](/en-us/dotnet/api/microsoft.aspnetcore.hosting.hostingenvironmentextensions.isdevelopment) returns true:
```

**Comment by JamesNK** [comment]
Applied

**Review by wadepickett — APPROVED** [review]
Looks good with the update.

**Pull request merged**